### DISCUSSION TITLE
After connecting via ssh contrib command, execute multiple commands in the same session

### DISCUSSION THREAD

**Discussion by sla-te** [amoffat/sh#701] [Q&A] [answered]
As of now, what I understood when using ssh is the following

```python
my_server = sh.ssh.bake("-i", "priv_key", "root@10.0.0.3")
print(my_server("ifconfig"))
print(my_server("whoami"))
```
This though will open two separate sessions for execution of each command, which makes it difficult to run a chain of commands within the same session. What would be the correct way to run multiple commands in the same session?

**Comment by amoffat** [comment]
Long running ssh sessions is not really what sh is designed for, though it can be done, with significant effort. You're better off using something like https://github.com/fabric/fabric

**Comment by sla-te** [comment]
Thank you for the suggestion, the amount of open issues in this repo scares me, what about using a context manager  as suggested here: https://github.com/amoffat/sh/issues/592?

**Reply by amoffat** [reply]
I had forgotten about that solution. It sounds like that could work. Try it out and share your results here and we can mark that as the answer if it works.

**Reply by sla-te** [reply]
In his explanation he describes to "keep the ssh session master alive" but as far as I understand sh syntax, there is only `_bg=True` which exits after the command is completed and `sh.ssh`executed with without any command in addition to perform on the server would exit right after connecting, would you by any chance have an elegant idea to handle this cleanly?

With my limited knowledge about sh my only idea would be to provide a `_out=interact` which does something like:

```python
while self.keep_alive:
    sleep(1)
 ```
and in `__exit__` setting `self.keep_alive` to False followed by `.wait()` on the process/command running in the background.

**Comment by sla-te** [comment, marked as answer]
This is what I ultimately came up with:

```python
from dataclasses import dataclass, field
from time import sleep
from typing import Any

import sh

@dataclass(slots=True)
class MasterSlaveSsh:
    master: Any = field(
        init=False,
        default_factory=lambda: sh.ssh.bake(
            "-o", "ControlMaster=yes", "-o", "ControlPath=~/.ssh/control-10.0.0.1_root", "-i", "priv_key", "root@10.0.0.1"
        ),
    )
    slave: Any = field(
        init=False,
        default_factory=lambda: sh.ssh.bake("-o", "ControlPath=~/.ssh/control-10.0.0.1_root", "-i", "priv_key", "root@10.0.0.1"),
    )

    def __enter__(self) -> slave:
        is_connected: bool = False

        def assert_connected(line):
            nonlocal is_connected

            if line.strip() == "sleeping...":
                is_connected = True

        self.master = self.master(
            'touch /root/heartbeat; while [ -f "/root/heartbeat" ]; do echo "sleeping..."; sleep 3; done;', _bg=True, _out=assert_connected
        )
        while not is_connected:
            sleep(1)

        print("Connected")

        return self.slave

    def __exit__(self, exc_type, exc_val, exc_tb):
        print(self.slave.rm("-v", "/root/heartbeat"))
        self.master.wait()


with MasterSlaveSsh() as ssh:
    print(ssh("echo 1 Doing work; sleep 3;"))
    print(ssh("echo 2 Doing work; sleep 3;"))
    print(ssh("echo 3 Doing work; sleep 3;"))
```

If you have any better ideas, do let me know, this way at least it uses the same connection - the shell/environment is still not the same per session though, maybe using [screen](https://linuxize.com/post/how-to-use-linux-screen/) would be most elegant for fixing that.

**Reply by amoffat** [reply]
Cool, thanks for sharing, this will be helpful to other users. If this solves your problem, go ahead and mark it as the answer :+1:
